diff --git a/adapter/outbound/socks5.go b/adapter/outbound/socks5.go
index f27f6de8..fd55fe84 100644
--- a/adapter/outbound/socks5.go
+++ b/adapter/outbound/socks5.go
@@ -2,12 +2,14 @@ package outbound
 
 import (
 	"context"
+	"encoding/binary"
 	"errors"
 	"fmt"
 	"io"
 	"net"
 	"net/netip"
 	"strconv"
+	"time"
 
 	N "github.com/metacubex/mihomo/common/net"
 	"github.com/metacubex/mihomo/component/ca"
@@ -25,6 +27,7 @@ type Socks5 struct {
 	tls            bool
 	skipCertVerify bool
 	tlsConfig      *tls.Config
+	udpOverTCP     bool
 }
 
 type Socks5Option struct {
@@ -36,6 +39,7 @@ type Socks5Option struct {
 	Password       string `proxy:"password,omitempty"`
 	TLS            bool   `proxy:"tls,omitempty"`
 	UDP            bool   `proxy:"udp,omitempty"`
+	UDPOverTCP     bool   `proxy:"udp-over-tcp,omitempty"`
 	SkipCertVerify bool   `proxy:"skip-cert-verify,omitempty"`
 	Fingerprint    string `proxy:"fingerprint,omitempty"`
 	Certificate    string `proxy:"certificate,omitempty"`
@@ -90,6 +94,41 @@ func (ss *Socks5) ListenPacketContext(ctx context.Context, metadata *C.Metadata)
 	if err = ss.ResolveUDP(ctx, metadata); err != nil {
 		return nil, err
 	}
+
+	if ss.udpOverTCP {
+		c, err := ss.dialer.DialContext(ctx, "tcp", ss.addr)
+		if err != nil {
+			return nil, fmt.Errorf("%s connect error: %w", ss.addr, err)
+		}
+
+		if ss.tls {
+			cc := tls.Client(c, ss.tlsConfig)
+			ctx, cancel := context.WithTimeout(context.Background(), C.DefaultTLSTimeout)
+			defer cancel()
+			err = cc.HandshakeContext(ctx)
+			c = cc
+			if err != nil {
+				c.Close()
+				return nil, err
+			}
+		}
+
+		var user *socks5.User
+		if ss.user != "" {
+			user = &socks5.User{
+				Username: ss.user,
+				Password: ss.pass,
+			}
+		}
+
+		if _, err := ss.clientHandshakeContext(ctx, c, serializesSocksAddr(metadata), socks5.CmdConnect, user); err != nil {
+			c.Close()
+			return nil, err
+		}
+
+		return newPacketConn(&udpOverTCPPacketConn{Conn: c}, ss), nil
+	}
+
 	c, err := ss.dialer.DialContext(ctx, "tcp", ss.addr)
 	if err != nil {
 		err = fmt.Errorf("%s connect error: %w", ss.addr, err)
@@ -205,6 +244,7 @@ func NewSocks5(option Socks5Option) (*Socks5, error) {
 		tls:            option.TLS,
 		skipCertVerify: option.SkipCertVerify,
 		tlsConfig:      tlsConfig,
+		udpOverTCP:     option.UDPOverTCP,
 	}
 	outbound.dialer = option.NewDialer(outbound.DialOptions())
 	return outbound, nil
@@ -248,3 +288,64 @@ func (uc *socksPacketConn) Close() error {
 	uc.tcpConn.Close()
 	return uc.PacketConn.Close()
 }
+
+type udpOverTCPPacketConn struct {
+	net.Conn
+}
+
+func (uc *udpOverTCPPacketConn) WriteTo(b []byte, addr net.Addr) (n int, err error) {
+	packet, err := socks5.EncodeUDPPacket(socks5.ParseAddrToSocksAddr(addr), b)
+	if err != nil {
+		return
+	}
+
+	header := make([]byte, 2)
+	binary.BigEndian.PutUint16(header, uint16(len(packet)))
+	if _, err = uc.Conn.Write(header); err != nil {
+		return
+	}
+	n, err = uc.Conn.Write(packet)
+	return len(b), err
+}
+
+func (uc *udpOverTCPPacketConn) ReadFrom(b []byte) (int, net.Addr, error) {
+	header := make([]byte, 2)
+	if _, err := io.ReadFull(uc.Conn, header); err != nil {
+		return 0, nil, err
+	}
+
+	length := binary.BigEndian.Uint16(header)
+	packet := make([]byte, length)
+	if _, err := io.ReadFull(uc.Conn, packet); err != nil {
+		return 0, nil, err
+	}
+
+	addr, payload, err := socks5.DecodeUDPPacket(packet)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	udpAddr := addr.UDPAddr()
+	if udpAddr == nil {
+		return 0, nil, errors.New("parse udp addr error")
+	}
+
+	copy(b, payload)
+	return len(payload), udpAddr, nil
+}
+
+func (uc *udpOverTCPPacketConn) LocalAddr() net.Addr {
+	return uc.Conn.LocalAddr()
+}
+
+func (uc *udpOverTCPPacketConn) SetDeadline(t time.Time) error {
+	return uc.Conn.SetDeadline(t)
+}
+
+func (uc *udpOverTCPPacketConn) SetReadDeadline(t time.Time) error {
+	return uc.Conn.SetReadDeadline(t)
+}
+
+func (uc *udpOverTCPPacketConn) SetWriteDeadline(t time.Time) error {
+	return uc.Conn.SetWriteDeadline(t)
+}
